backup: set details.DownloadSpans in blocking OR job

msbutler opened the pull request:
As soon as the link phase starts, OR must be prepared to excise the restoring
key space in the event of a job failure or cancellation. This patch persists
DownloadSpans to the job record so the OnFailOrCancel routine can easily begin
recovery. This patch also slightly refactors the doDownloadSpans args to always
use details.DownloadSpans.

Epic: none

Release note: none

msbutler committed:
backup: set details.DownloadSpans before link phase starts

As soon as the link phase starts, OR must be prepared to excise the restoring
key space in the event of a job failure or cancellation. This patch persists
DownloadSpans to the job record so the OnFailOrCancel routine can easily begin
recovery. This patch also slightly refactors the doDownloadSpans args to always
use details.DownloadSpans.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144048)
<!-- Reviewable:end -->

jeffswenson commented:
Is there a specific bug this is intended to fix or are is this a preventative measure?

jeffswenson commented on pkg/backup/restore_job.go:
```diff
@@ -1816,6 +1816,21 @@ func (r *restoreResumer) doResume(ctx context.Context, execCtx interface{}) erro
+		// Ensure we have the latest copy of the job details.
+		details = r.job.Details().(jobspb.RestoreDetails)
+		details.DownloadSpans = downloadSpans
+		if err := r.job.NoTxn().SetDetails(ctx, details); err != nil {
```
Are there any concurrent processes that might be writing to details after the job starts? I'm wondering if we have to worry about overwriting a some change that committed between when the job started and when we checkpoint the details.

msbutler replied:
I am not, as the restore flow has not begun yet. While it is sad that we have to update job details after job planning, this felt like the change with the lowest blast radius. In the func above (`createImportingDescriptors`) we also update the job details. 

I think this is safe so long as a refresh my local copy of `details`. All updates to details will also update the in memory copy in `r.job.Details()`

jeffswenson replied:
`createImportingDescriptors` is using a txn, but you are right that it doesn't actually read the details in the transaction. So at the very least this isn't introducing a new bug. 

In general I would prefer if we read the details, update, and write them in a transaction so that we don't need to think carefully about what code paths might be updating details. But if this is common practice for jobs 🤷 As long as we are consistent with the convention and only the job is allowed to update the details outside of a txn.

msbutler replied:
big yak -> refactoring how jobs update their state

msbutler committed:
backup: add tenant spans in createImportingDescriptors

We should treat the output of createImportingDescriptors as static to reduce
probability of bugs. In a future PR, we should split createImportingDescriptors
into 2 functions: one that commits the importing descriptors and another that
creates the target restore spans.

Epic: none

Release note: none

msbutler commented:
> Is there a specific bug this is intended to fix or are is this a preventative measure?
Added more context to the commit msg.

jeffswenson reviewed (approved):
LGTM

msbutler committed:
backup: a restore with no table data but with types should exit early

Epic: none

Release note: none

jeffswenson reviewed (approved):
LGTM

msbutler commented:
TFTR!

bors r=jeffswenson

msbutler commented:
blathers backport 25.2
